src/sctools/bam.py

```diff
@@ -0,0 +1,125 @@
+import pysam
+
+"""
+unlike fastq and gtf which lack flexible iterators, the pysam wrapper provides an excellent iterator 
+over sam records that would be difficult to improve upon in pure python. 
+
+This module provides a few additional methods that extend the functionality of pysam
+"""
+
+
+class SubsetAlignments:
+    """Wrapper for pysam/htslib that enables non-standard filtering of alignments in a bamfile
+
+    :method indices_by_chromosome: extract reads specific to a chromosome
+    """
+
+    def __init__(self, alignment_file, open_mode=None):
+        """
+        :param str alignment_file: sam or bam file.
+        :param str open_mode: optional, mode to read file. Will be autodetected by file type if
+          the file contains the correct suffix for its type.
+        """
+
+        if open_mode is None:
+            if alignment_file.endswith('.bam'):
+                open_mode = 'rb'
+            elif alignment_file.endswith('.sam'):
+                open_mode = 'r'
+            else:
+                raise ValueError('could not autodetect file type for alignment_file %s '
+                                 '(detectible suffixes: .sam, .bam)' % alignment_file)
+        self._file = alignment_file
+        self._open_mode = open_mode
+
+    def indices_by_chromosome(self, n_specific, chromosome, include_other=0):
+        """Return the list of first n_specific indices of reads aligned to selected chromosome.
+
+        If desired, will also return non-specific indices in a second list (can serve as negative
+        control reads).
+
+        :param int n_specific: number of aligned reads to return indices for
+        :param str chromosome: only reads from this chromosome are considered valid
+        :param int include_other: optional (default=0), the number of unaligned reads
+          to include in the file.
+        :return [int]: list of aligned indicies
+        :return [int]: list of unaligned indices, only returned if include_other is not zero.
+        """
+
+        # check chromosome
+        valid_chromosomes = [str(i) for i in range(1, 23)] + ['M', 'X', 'Y']
+        if isinstance(chromosome, int) and chromosome < 23:
+            chromosome = str(chromosome)  # try to convert
+        if chromosome not in valid_chromosomes:
+            raise ValueError('chromsome %s not valid. Must be one of %r' %
+                             (chromosome, valid_chromosomes))
+
+        with pysam.AlignmentFile(self._file, self._open_mode) as fin:
+            specific, nonspecific = 0, 0  # counters
+            chromosome = str(chromosome)
+            chromosome_indices = []
+            other_indices = []
+
+            for i, record in enumerate(fin):
```

Comment:
It feels like some of the contents here, especially the if else block would be worth moving out into its own function that could be unit tested on its own.

Reply:
I simplified the code. Since each list hashes its length, I don't need the counters; I can just test the length of `chromosome_indices` and `other_indices` directly. 

However, removing this into its own function would only serve to separate the type/value checking from the index extraction; I think this belongs where it is, but hope that the code simplification makes it clearer and reduces your desire to extract the code.